File: go/internal/endpoint/preflight.go
package endpoint

import (
	"cmp"
	"crypto/rand"
	"encoding/hex"
	"encoding/json/v2"
	"net"
	"net/http"
	"net/url"
	"strings"

	"github.com/zR-JB/graphite-meter/go/internal/config"
	"github.com/zR-JB/graphite-meter/go/internal/origin"
	"github.com/zR-JB/graphite-meter/go/internal/wire"
)

// Preflight advertises the measurement targets a client should use. generation is a per-process random tag.
type Preflight struct {
	cfg        *config.Config
	generation string
}

// NewPreflight builds the preflight endpoint for cfg with a fresh generation tag.
func NewPreflight(cfg *config.Config) *Preflight {
	var id [16]byte
	if _, err := rand.Read(id[:]); err != nil {
		panic(err)
	}
	return &Preflight{cfg: cfg, generation: hex.EncodeToString(id[:])}
}

func (p *Preflight) HandleHTTP(w http.ResponseWriter, r *http.Request) error {
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-store")
	return json.MarshalWrite(w, p.build(r))
}

func (p *Preflight) build(r *http.Request) wire.Preflight {
	// SplitHostPort fails on a port-less Host; the fallback also unwraps the brackets of a literal IPv6 authority.
	host, _, _ := net.SplitHostPort(r.Host)
	host = cmp.Or(host, strings.TrimPrefix(strings.TrimSuffix(r.Host, "]"), "["))
	return p.buildForHost(host)
}

// ConnectOrigins lists distinct cross-origin targets for host.
func (p *Preflight) ConnectOrigins(host string) []string {
	pf := p.buildForHost(host)
	seen := map[string]bool{"": true, ".": true}
	out := make([]string, 0)
	add := func(o string) {
		if !seen[o] {
			seen[o] = true
			out = append(out, o)
		}
	}
	for _, t := range pf.Capabilities.ThroughputTargets {
		add(t.Origin)
	}
	for _, t := range pf.Capabilities.LatencyTargets {
		add(t.Origin)
		add(websocketOrigin(t.Origin))
	}
	return out
}

func websocketOrigin(target string) string {
	if host, ok := strings.CutPrefix(target, "https://"); ok {
		return "wss://" + host
	}
	if host, ok := strings.CutPrefix(target, "http://"); ok {
		return "ws://" + host
	}
	return ""
}

func (p *Preflight) buildForHost(host string) wire.Preflight {
	throughput := make([]wire.ThroughputTarget, 0)
	latency := make([]wire.LatencyTarget, 0)
	addThroughput := func(base, protocol string) {
		base = strings.TrimRight(base, "/")
		for i := range throughput {
			if throughput[i].Transport == wire.TransportFetchStream && origin.Equal(throughput[i].Origin, base) {
				if throughput[i].Protocol != protocol {
					throughput[i].Protocol = "negotiated"
				}
				return
			}
		}
		throughput = append(throughput, wire.ThroughputTarget{ID: base, Origin: base, Transport: wire.TransportFetchStream, Protocol: protocol, TLS: strings.HasPrefix(base, "https://"), Routes: wire.DefaultThroughputRoutes()})
	}
	addLatency := func(base string) {
		base = strings.TrimRight(base, "/")
		for _, e := range latency {
			if e.Transport == wire.TransportWebSocket && origin.Equal(e.Origin, base) {
				return
			}
		}
		latency = append(latency, wire.LatencyTarget{ID: base, Origin: base, Transport: wire.TransportWebSocket, Protocol: "http1", TLS: strings.HasPrefix(base, "https://"), Routes: wire.DefaultLatencyRoutes()})
	}
	addWebTransport := func(base string) {
		base = strings.TrimRight(base, "/")
		throughput = append(throughput, wire.ThroughputTarget{ID: base, Origin: base, Transport: wire.TransportWebTransport, Protocol: "http3", TLS: true, Routes: wire.DefaultThroughputRoutes()})
		throughput = append(throughput, wire.ThroughputTarget{ID: base, Origin: base, Transport: wire.TransportWebTransportDatagram, Protocol: "http3", TLS: true, Routes: wire.DefaultThroughputRoutes()})
		latency = append(latency, wire.LatencyTarget{ID: base, Origin: base, Transport: wire.TransportWebTransport, Protocol: "http3", TLS: true, Routes: wire.DefaultLatencyRoutes()})
	}
	native := []struct {
		name, public, scheme, addr, protocol string
		latency, webTransport                bool
	}{
		{config.NativeH1Clear, p.cfg.NativePublic.H1, "http", p.cfg.Native.H1, "http1", true, false},
		{config.NativeH1TLS, p.cfg.NativePublic.H1TLS, "https", p.cfg.Native.H1TLS, "http1", true, false},
		{config.NativeH2, p.cfg.NativePublic.H2, "https", p.cfg.Native.H2, "http2", false, false},
		{config.NativeH3, p.cfg.NativePublic.H3, "https", p.cfg.Native.H3, "http3", false, true},
	}
	for _, e := range native {
		if p.cfg.NativeAdvertised(e.name) {
			base := nativeOrigin(e.public, e.scheme, host, e.addr)
			addThroughput(base, e.protocol)
			if e.latency {
				addLatency(base)
			}
			if e.webTransport {
				addWebTransport(base)
			}
		}
	}
	for _, base := range p.cfg.Public.Both {
		addThroughput(publicBase(base), "negotiated")
		addLatency(publicBase(base))
	}
	for _, base := range p.cfg.Public.Throughput {
		addThroughput(publicBase(base), "negotiated")
	}
	for _, base := range p.cfg.Public.Latency {
		addLatency(publicBase(base))
	}
	return wire.Preflight{Server: wire.ServerInfo{Name: p.cfg.ServerName, Location: p.cfg.ServerLocation}, EngineVersion: p.cfg.EngineVersion, Generation: p.generation, Capabilities: wire.Capabilities{UploadCheckpoint: true, ThroughputTargets: throughput, LatencyTargets: latency}}
}

func publicBase(configured string) string {
	if configured == "self" {
		return "."
	}
	return configured
}

func nativeOrigin(public, scheme, host, addr string) string {
	if public != "" {
		return public
	}
	_, port, err := net.SplitHostPort(addr)
	if err != nil || port == "" {
		return scheme + "://" + host
	}
	return (&url.URL{Scheme: scheme, Host: net.JoinHostPort(host, port)}).String()
}

File: rust/server/src/preflight.rs
//! Discovery is built once from resolved policy, independent of the listener adapter.
use crate::config::{Config, ConfigError, NativeKind};
use graphite_meter_core::{
    discovery::{
        Capabilities, LatencyTarget, LatencyTransport, Preflight as Document, Protocol, ServerInfo,
        ThroughputTarget, ThroughputTransport,
    },
    origin::{key, target_origin},
};
use std::sync::Arc;

pub struct Preflight {
    config: Arc<Config>,
    generation: String,
}

impl Preflight {
    pub fn new(config: Arc<Config>) -> Result<Self, ConfigError> {
        config.validate()?;
        let mut nonce = [0_u8; 16];
        rustls::crypto::ring::default_provider()
            .secure_random
            .fill(&mut nonce)
            .map_err(|_| "failed to generate discovery identity")?;
        let generation = nonce.iter().map(|byte| format!("{byte:02x}")).collect();
        Ok(Self { config, generation })
    }

    pub fn build(&self, authority: &str) -> Result<Document, ConfigError> {
        if authority.contains('@') {
            return Err("request authority must not contain credentials".into());
        }
        let authority: http::uri::Authority = authority.parse()?;
        let host = authority
            .host()
            .trim_start_matches('[')
            .trim_end_matches(']');
        self.build_for_host(host)
    }

    pub fn build_for_host(&self, host: &str) -> Result<Document, ConfigError> {
        let config = &self.config;
        let mut capabilities = Capabilities {
            upload_checkpoint: true,
            throughput: Vec::new(),
            latency: Vec::new(),
        };
        for kind in NativeKind::ALL {
            if !config.native_advertised(kind) {
                continue;
            }
            let listener = config.listener(kind);
            let base = if listener.public_origin.is_empty() {
                native_origin(kind, host, &listener.address)?
            } else {
                listener.public_origin.clone()
            };
            let protocol = match kind {
                NativeKind::H1 | NativeKind::H1Tls => Protocol::Http1,
                NativeKind::H2 => Protocol::Http2,
                NativeKind::H3 => Protocol::Http3,
            };
            add_throughput(&mut capabilities, &base, protocol);
            if matches!(kind, NativeKind::H1 | NativeKind::H1Tls) {
                add_latency(&mut capabilities, &base);
            }
            if kind == NativeKind::H3 {
                for transport in [
                    ThroughputTransport::WebTransport,
                    ThroughputTransport::WebTransportDatagram,
                ] {
                    capabilities.throughput.push(ThroughputTarget {
                        base_url: base.clone(),
                        transport,
                        protocol: Protocol::Http3,
                    });
                }
                capabilities.latency.push(LatencyTarget {
                    base_url: base,
                    transport: LatencyTransport::WebTransport,
                });
            }
        }
        for raw in &config.public.both {
            add_throughput(&mut capabilities, public_base(raw), Protocol::Negotiated);
            add_latency(&mut capabilities, public_base(raw));
        }
        for raw in &config.public.throughput {
            add_throughput(&mut capabilities, public_base(raw), Protocol::Negotiated);
        }
        for raw in &config.public.latency {
            add_latency(&mut capabilities, public_base(raw));
        }
        let document = Document {
            server: ServerInfo {
                name: config.server_name.clone(),
                location: config.server_location.clone(),
            },
            engine_version: config.engine_version.clone(),
            generation: self.generation.clone(),
            capabilities,
        };
        document.validate()?;
        Ok(document)
    }

    pub fn connect_origins(&self, host: &str) -> Result<Vec<String>, ConfigError> {
        let document = self.build_for_host(host)?;
        let mut origins = Vec::new();
        let mut add = |origin: String| {
            if !origin.is_empty() && origin != "." && !origins.contains(&origin) {
                origins.push(origin);
            }
        };
        for target in document.capabilities.throughput {
            add(target.base_url);
        }
        for target in document.capabilities.latency {
            add(target.base_url.clone());
            if let Some(host) = target.base_url.strip_prefix("https://") {
                add(format!("wss://{host}"));
            } else if let Some(host) = target.base_url.strip_prefix("http://") {
                add(format!("ws://{host}"));
            }
        }
        Ok(origins)
    }
}

fn add_throughput(capabilities: &mut Capabilities, base: &str, protocol: Protocol) {
    let base = base.trim_end_matches('/');
    if let Some(target) = capabilities.throughput.iter_mut().find(|target| {
        target.transport == ThroughputTransport::FetchStream && key(&target.base_url) == key(base)
    }) {
        if target.protocol != protocol {
            target.protocol = Protocol::Negotiated;
        }
    } else {
        capabilities.throughput.push(ThroughputTarget {
            base_url: base.into(),
            transport: ThroughputTransport::FetchStream,
            protocol,
        });
    }
}

fn add_latency(capabilities: &mut Capabilities, base: &str) {
    let base = base.trim_end_matches('/');
    if !capabilities.latency.iter().any(|target| {
        target.transport == LatencyTransport::WebSocket && key(&target.base_url) == key(base)
    }) {
        capabilities.latency.push(LatencyTarget {
            base_url: base.into(),
            transport: LatencyTransport::WebSocket,
        });
    }
}

fn public_base(raw: &str) -> &str {
    if raw == "self" { "." } else { raw }
}

fn native_origin(kind: NativeKind, host: &str, address: &str) -> Result<String, ConfigError> {
    let scheme = if kind == NativeKind::H1 {
        "http"
    } else {
        "https"
    };
    let host = if host.contains(':') {
        format!("[{host}]")
    } else {
        host.to_owned()
    };
    let port = address
        .rsplit_once(':')
        .map(|(_, port)| port)
        .filter(|port| !port.is_empty());
    let origin = match port {
        Some(port) => format!("{scheme}://{host}:{port}"),
        None => format!("{scheme}://{host}"),
    };
    target_origin(&origin)?;
    Ok(origin)
}
